platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/GraphVariantSelector.java

```diff
@@ -80,7 +80,7 @@ public VariantGraphResolveState selectByAttributeMatching(
             ComponentGraphResolveMetadata targetComponent = targetComponentState.getMetadata();
             AttributeMatcher attributeMatcher = consumerSchema.withProducer(targetComponent.getAttributesSchema());
             GraphSelectionCandidates candidates = targetComponentState.getCandidatesForGraphVariantSelection();
-            throw failureHandler.noMatchingGraphVariantFailure(consumerSchema, attributeMatcher, consumerAttributes, targetComponent, candidates);
+            throw failureHandler.noCompatibleVariantsFailure(consumerSchema, attributeMatcher, targetComponentState, consumerAttributes, ImmutableSet.copyOf(explicitRequestedCapabilities), candidates);
```

Comment:
`ImmutableSet.copyOf(explicitRequestedCapabilities)` is counterintuitive to the existence of `ImmutableCapabilities`. 

Ideally, `explicitlyRequestedCapabilities` would _already_ be an `ImmutableCapabilities` instance and we would not need to wrap it here. We _should_ make it an `ImmutableCapabilities`, but not in this PR, as I don't want to expand the scope anymore. 

However, I would much prefer `ImmutableCapabilities.of(explicitRequestedCapabilities)` instead and to use that type throughout the resolution failure infrastructure.

Reply:
Applies throughout this file

Reply:
Looking more in-depth at the `ResoluionFailureHandler`, it is clear we are mixing types. Some methods accept `ImmutableCapabilities`, while others accept `Set<Capability>`. We should use `ImmutableCapabilities` everywhere.

Reply:
Yeah, we should be consistent.  At least until/unless we ever decide to get rid of `ImmutableCapabiliites`.  I'll update these.